### PR TITLE
ET-860: Add trigger base system

### PR DESCRIPTION
**Pull request by mitogh**
Add support for trigger base messages, this phase of the process is only associated with attendees actions for RSVP and tribe commerce. 

RSVP:

- Event Checkin
- Event Going
- Event Not Going

Tribe Commerce: 

- Ticket generated.

Ticket: [ET-860]

Artifact: http://p.tri.be/j1XzyG

[ET-860]: https://moderntribe.atlassian.net/browse/ET-860

### CODE REVIEW

File: src/Tribe/Promoter/Triggers/Factory.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	 * When an action `tribe_tickets_promoter_trigger_attendee` is fired, react with an attendee trigger.
+	 *
+	 * @since TBD
+	 *
+	 * @param string                  $type     The type of trigger message.
+	 * @param AttendeeModel           $attendee The representation of the attendee.
+	 * @param Tribe__Tickets__Tickets $ticket   The ticket provider instance.
+	 */
+	public function build_attendee( $type, $attendee, $ticket ) {
+		do_action( 'tribe_tickets_promoter_trigger', new AttendeeTrigger( $type, $attendee, $ticket ) );
```
Comment (bordoni) [inline src/Tribe/Promoter/Triggers/Factory.php:31]: Needs docblock.

File: src/Tribe/Promoter/Triggers/Observers/Commerce.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	 * @param int    $product_id            PayPal ticket post ID
+	 * @param int    $order_attendee_id     Attendee number in submitted order
+	 * @param string $attendee_order_status The order status for the attendee.
+	 */
+	public function attendee_created( $attendee_id, $order_id, $product_id, $order_attendee_id, $attendee_order_status ) {
+		/** @var Tribe__Tickets__Commerce__PayPal__Main $ticket */
+		$ticket   = tribe( 'Tribe__Tickets__Commerce__PayPal__Main' );
+		$attendee = new Attendee( $ticket->get_attendee( $attendee_id ) );
+
+		do_action( 'tribe_tickets_promoter_trigger_attendee', 'ticket_purchased', $attendee, $ticket );
```
Comment (bordoni) [inline src/Tribe/Promoter/Triggers/Observers/Commerce.php:35]: Needs docblock.

File: src/Tribe/Promoter/Triggers/Observers/RSVP.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	 *
+	 * @param $type
+	 * @param $attendee_id
+	 */
+	private function trigger( $type, $attendee_id ) {
+		/** @var Tribe__Tickets__RSVP $ticket */
+		$ticket   = tribe( 'tickets.rsvp' );
+		$attendee = new Attendee( $ticket->get_attendee( $attendee_id ) );
+
+		do_action( 'tribe_tickets_promoter_trigger_attendee', $type, $attendee, $ticket );
```
Comment (bordoni) [inline src/Tribe/Promoter/Triggers/Observers/RSVP.php:75]: Needs docblocks.

File: src/Tribe/Promoter/Triggers/Builders/AttendeeTrigger.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	private $type;
+
+	/**
+	 * Attendee constructor.
+	 *
+	 * @param string                  $trigger_name
+	 * @param Model                   $attendee
+	 * @param Tribe__Tickets__Tickets $provider
+	 */
+	public function __construct( $trigger_name, $attendee, $provider ) {
```
Comment (bordoni) [inline src/Tribe/Promoter/Triggers/Builders/AttendeeTrigger.php:33]: Type hint.

File: src/Tribe/Promoter/Service_Provider.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ */
+class Service_Provider extends tad_DI52_ServiceProvider {
+
+	/**
+	 * Binds and sets up implementations.
+	 *
+	 * @since TBD
+	 */
+	public function register() {
+		$this->container->singleton( 'tickets.promoter.integration', Tribe__Tickets__Promoter__Integration::class, [ 'hook' ] );
```
Comment (bordoni) [inline src/Tribe/Promoter/Service_Provider.php:26]:
```suggestion
		$this->container->singleton( 'tickets.promoter.integration', Tribe__Tickets__Promoter__Integration::class, [ 'hook' ] );
		$this->container->singleton( Tribe__Tickets__Promoter__Integration::class, Tribe__Tickets__Promoter__Integration::class, [ 'hook' ] );
```

File: src/Tribe/Promoter/Triggers/Dispatcher.php [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	/**
+	 * @var Tribe__Promoter__Connector
+	 */
+	private $connector;
+	/**
+	 * @var Tribe__Promoter__PUE
+	 */
+	private $pue;
+
+	public function __construct( Tribe__Promoter__Connector $connector, Tribe__Promoter__PUE $pue ) {
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Dispatcher.php:29]:
```suggestion
	/**
	 * Setup the dispatcher.
	 *
	 * @param Tribe__Promoter__Connector $connect The connector object for Promoter.
	 * @param Tribe__Promoter__PUE       $pue     The PUE object for Promoter.
	 */
	public function __construct( Tribe__Promoter__Connector $connector, Tribe__Promoter__PUE $pue ) {
```

File: src/Tribe/Promoter/Triggers/Dispatcher.php [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	/**
+	 * Creat ea payload using the trigger object.
+	 *
+	 * @since TBD
+	 *
+	 * @param Triggered $trigger The trigger object creating this action.
+	 *
+	 * @return array The payload to be encoded and delivered to the connector app.
+	 */
+	private function get_payload( $trigger ) {
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Dispatcher.php:95]:
```suggestion
	private function get_payload( Triggered $trigger ) {
```

File: src/Tribe/Promoter/Triggers/Factory.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	/**
+	 * When an action `tribe_tickets_promoter_trigger_attendee` is fired, react with an attendee trigger.
+	 *
+	 * @since TBD
+	 *
+	 * @param string                  $type     The type of trigger message.
+	 * @param AttendeeModel           $attendee The representation of the attendee.
+	 * @param Tribe__Tickets__Tickets $ticket   The ticket provider instance.
+	 */
+	public function build_attendee( $type, $attendee, $ticket ) {
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Factory.php:30]:
```suggestion
	public function build_attendee( $type, AttendeeModel $attendee, Tribe__Tickets__Tickets $ticket ) {
```

File: readme.txt [hunk 9]
```diff
@@ -118,6 +118,10 @@ Currently, the following add-ons are available for Event Tickets:
 
 == Changelog ==
 
+= [] TBD
+
+* Feature - Add foundation to support trigger base messages on Promoter [ET-860]
```
Comment (sc0ttkclark) [inline readme.txt:123]:
```suggestion
* Feature - Add foundation to support trigger base messages on Promoter. [ET-860]
```

File: src/Tribe/Promoter/Triggers/Observers/Commerce.php [hunk 10]
```diff
@@ -28,10 +31,43 @@ public function hook() {
...
+	public function checkin( $attendee_id, $qr ) {
+		$this->trigger( 'checkin', $attendee_id );
+	}
+
+	/**
+	 * Fire an trigger action using the tribe commerce as main source of the ticket data.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $type        The trigger type
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Observers/Commerce.php:54]:
```suggestion
	 * @param string $type        The trigger type.
```

File: src/Tribe/Promoter/Triggers/Observers/Commerce.php [hunk 11]
```diff
@@ -28,10 +31,43 @@ public function hook() {
 	 * @param string $attendee_order_status The order status for the attendee.
 	 */
 	public function attendee_created( $attendee_id, $order_id, $product_id, $order_attendee_id, $attendee_order_status ) {
+		$this->trigger( 'ticket_purchased', $attendee_id );
+	}
+
+	/**
+	 * Responds to a checkin action
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Observers/Commerce.php:38]:
```suggestion
	 * Responds to a checkin action.
```

File: src/Tribe/Promoter/Triggers/Observers/Commerce.php [hunk 12]
```diff
@@ -28,10 +31,43 @@ public function hook() {
...
+	 *
+	 * @since TBD
+	 *
+	 * @param int       $attendee_id
+	 * @param bool|null $qr
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Observers/Commerce.php:43]:
```suggestion
	 * @param int       $attendee_id The ID of the attendee utilized.
	 * @param bool|null $qr          Whether it's from a QR scan.
```

File: src/Tribe/Promoter/Triggers/Observers/RSVP.php [hunk 13]
```diff
@@ -18,12 +20,18 @@ class RSVP {
...
 		add_action( 'event_tickets_rsvp_attendee_created', [ $this, 'attendee_created' ], 10, 3 );
 		add_action( 'updated_postmeta', [ $this, 'attendee_updated' ], 10, 4 );
 	}
 
-	public function rsvp_checkin( $attendee_id ) {
+	/**
+	 * Fires a checkin action
+	 *
+	 * @var int       $attendee_id
+	 * @var bool|null $qr
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Observers/RSVP.php:32]:
```suggestion
	 * @param int       $attendee_id The ID of the attendee utilized.
	 * @param bool|null $qr          Whether it's from a QR scan.
```

File: src/Tribe/Promoter/Triggers/Observers/RSVP.php [hunk 14]
```diff
@@ -18,12 +20,18 @@ class RSVP {
...
 	public function hook() {
-		add_action( 'rsvp_checkin', [ $this, 'rsvp_checkin' ] );
+		add_action( 'rsvp_checkin', [ $this, 'rsvp_checkin' ], 10, 2 );
 		add_action( 'event_tickets_rsvp_attendee_created', [ $this, 'attendee_created' ], 10, 3 );
 		add_action( 'updated_postmeta', [ $this, 'attendee_updated' ], 10, 4 );
 	}
 
-	public function rsvp_checkin( $attendee_id ) {
+	/**
+	 * Fires a checkin action
```
Comment (sc0ttkclark) [inline src/Tribe/Promoter/Triggers/Observers/RSVP.php:29]:
```suggestion
	 * Fires a checkin action.
```

**Comment by mitogh** [comment]
Thanks @sc0ttkclark let me know if you have additional suggestions.

**Pull request merged**